universal-application-tool-0.0.1/app/repository/ApplicationRepository.java

```diff
@@ -5,62 +5,169 @@
+              .eq("program.name", program.getProgramDefinition().name())
+              .findList();
+      for (Application application : oldApplications) {
+        // Delete any in-progress drafts, and mark obsolete any old applications.
+        if (application.getLifecycleStage().equals(LifecycleStage.DRAFT)) {
+          application.setLifecycleStage(LifecycleStage.DELETED);
+        } else {
+          application.setLifecycleStage(LifecycleStage.OBSOLETE);
+        }
+        application.save();
+      }
+      Application application = new Application(applicant, program, LifecycleStage.ACTIVE);
+      application.save();
+      ebeanServer.commitTransaction();
+      return application;
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
+  private CompletionStage<Optional<Application>> perform(
+      long applicantId, long programId, Function<ApplicationArguments, Application> fn) {
     CompletionStage<Optional<Applicant>> applicantDb =
         applicantRepository.lookupApplicant(applicantId);
     CompletionStage<Optional<Program>> programDb = programRepository.lookupProgram(programId);
-    return applicantDb.thenCombineAsync(
-        programDb,
-        (applicantMaybe, programMaybe) -> {
-          if (applicantMaybe.isEmpty() || programMaybe.isEmpty()) {
-            return Optional.empty();
-          }
-          Applicant applicant = applicantMaybe.get();
-          Program program = programMaybe.get();
-          Application application = new Application(applicant, program, clock.instant());
-          application.save();
-          return Optional.of(application);
-        });
+    return applicantDb
+        .thenCombineAsync(
+            programDb,
+            (applicantMaybe, programMaybe) -> {
+              if (applicantMaybe.isEmpty()) {
+                throw new RuntimeException(new ApplicantNotFoundException(applicantId));
+              }
+              if (programMaybe.isEmpty()) {
+                throw new RuntimeException(new ProgramNotFoundException(programId));
+              }
+              return new ApplicationArguments(programMaybe.get(), applicantMaybe.get());
+            })
+        .thenApplyAsync(fn)
+        .thenApplyAsync(application -> Optional.of(application))
+        .exceptionally(
+            exception -> {
+              LOG.error(exception.toString());
+              exception.printStackTrace();
+              return Optional.empty();
+            });
+  }
+
+  // Need to transmit both arguments to submitApplication through the CompletionStage pipeline.
+  // Not useful in the API, not needed more broadly.
+  private static class ApplicationArguments {
+    public Program program;
+    public Applicant applicant;
+
+    public ApplicationArguments(Program program, Applicant applicant) {
+      this.program = program;
+      this.applicant = applicant;
+    }
+  }
+
+  private Application createOrUpdateDraftApplicationInternal(Applicant applicant, Program program) {
```

Comment:
Suggestion (blocking): does this actually update the application if it already exists? I'd expect a method called `update...`  to take args specifying what attributes to update. Suggest renaming this and relatedly named methods to `ensureDraftApplicationInternal`

Reply:
I think it does, yeah - I'll add a test to make sure it does.  The only thing to update is submission time, though.  Applicant and program are the args specifying which entry to update - they uniquely identify a draft.